rpc/lib/server/handlers.go

```diff
@@ -558,15 +542,33 @@ func (wsc *wsConnection) readRoutine() {
 
 // receives on a write channel and writes out on the socket
 func (wsc *wsConnection) writeRoutine() {
-	defer wsc.baseConn.Close()
+	pingTicker := time.NewTicker(wsc.pingPeriod)
+	defer func() {
+		pingTicker.Stop()
+		wsc.baseConn.Close()
+	}()
+
+	// https://github.com/gorilla/websocket/issues/97
+	pongs := make(chan string, 1)
+	wsc.baseConn.SetPingHandler(func(m string) error {
+		select {
```

Comment:
> How does this actually cause a deadlock if we're just using WriteControl in the PingHandler?

I think a deadlock is a wrong word here because it can lock up to `write timeout`. So it is not a deadlock per se, just temporary lock.